### ISSUE TITLE
iOS SDK starts location services on launch - need a way to delay/defer it

### ISSUE THREAD

**Issue by cowgp** [mapbox/mapbox-gl-native#4259]
A recent update to the iOS SDK added this line:
https://github.com/mapbox/mapbox-gl-native/blob/master/platform/ios/src/MGLMapboxEvents.m#L287

The the init of MGLMapboxEvents calls
     
     _paused = YES;
      [self resumeMetricsCollection];

but this actually makes it pass through and trigger the newly added line 287 linked above and actually initialize and start tracking user location during applicationDidFinishLaunching

This would be fine if my app was like the demo and I wanted the first thing you see to be a map, but map features are just one aspect of my app, I'd like to thus be in control and not start tracking a user till they have actually opened a view with a map on it.

**Comment by friedbunny** [comment]
Thanks for your awesome feedback, this sounds like intended behavior — `MGLMapboxEvents` is the iOS SDK's telemetry class.

Lately, we’ve been doing a ton of work to improve telemetry and the user experience around that feature. There [shouldn’t have been any changes](https://github.com/mapbox/mapbox-gl-native/commit/09c91008910746667b26d04eca679dfa5bb6b03b) to when telemetry will run while an app is in the foreground — it starts when the app starts and isn’t tied to showing the user location on a map.

You’ve already poked through the source, but also check out [our telemetry overview](https://www.mapbox.com/telemetry/) for more information on how we use this aggregated data (_spoiler: to improve OSM_) and about how fiercely we keep it anonymous and private. We're happy to answer questions you might have about any of this here, or you're always welcome to get in touch with us [through our website](https://www.mapbox.com/contact/).

**Comment by cowgp** [comment]
The issue is the user experience for user who's not using the map based features of our app - They see the compass icon in status bar and then when backgrounding the app are presented with the blue status bar "<App name> is using your location" Granted, it goes away, but still it raises the question by the user, "wait, why was that app using my location?"  

Currently, neither MGLMapboxEvents.h nor MGLLocationManager.h are exposed in the API so I have no ability to call stopUpdatingLocation on anything to manually stop it should I want to do that.   I understand why you want telemetry data - but I still feel strongly that there should be an opt out mechanism for developers who only want/need maps on screens that are not core functionality.  The current stance is heavy handed and assumes that anybody who is adding mapbox SDK has maps front and center as primary purpose of the app.

**Comment by postmechanical** [comment]
I support @cowgp's position. Individual apps should be responsible for requesting permission, starting, and stopping location services.

**Comment by postmechanical** [comment]
This is still a problem. Setting the access token should not immediately attempt to enable location permissions because the telemetry managing class needs to init. Telemetry should only begin accessing user location when a map view instance is actually rendered.

Also, there appears to be a bug here: https://github.com/mapbox/mapbox-gl-native/blob/5eda74a514964d1cac684483bafa08d458175f9a/platform/ios/src/MGLMapboxEvents.m#L281
It seems that it should rather be `if (self.paused` rather than `if (!self.paused`.

**Comment by friedbunny** [comment]
@postmechanical That `if !paused` is there to prevent multiple calls to `resumeMetricsCollection` from reinitializing `MGLMapboxEventsData` and re-starting location services unnecessarily — it is the “unpause” method, so it only needs to fully run if telemetry is paused.

Also, if anyone is seeing telemetry itself ask for initial location permission (independent of you asking for it by setting `MGLMapView.showsUserLocation`), that’ll be a regression that needs to be fixed.

**Comment by boundsj** [comment]
@friedbunny is correct about `if !paused` being there to prevent multiple calls to `resumeMetricsCollection`. The guard is really there to prevent multiple calls to `self.locationManager startUpdatingLocation` at the bottom of `resumeMetricsCollection`. Calling `startUpdatingLocation` triggers location event updates from iOS and we don't want to do that unnecessarily. That said, the way the internal state management is written can be a little hard to follow and additional refactoring is in order.

> Also, if anyone is seeing telemetry itself ask for initial location permission (independent of you asking for it by setting MGLMapView.showsUserLocation), that’ll be a regression that needs to be fixed.

Telemetry itself will [never ask for location permission](https://github.com/mapbox/mapbox-gl-native/blob/5eda74a514964d1cac684483bafa08d458175f9a/platform/ios/src/MGLLocationManager.m#L64-L65)

**Comment by boundsj** [comment]
> @postmechanical  Setting the access token should not immediately attempt to enable location permissions because the telemetry managing class needs to init.

The location permissions prompt that a user sees will only be triggered by the Mapbox SDK when the map view's `showsUserLocation` is sent YES. Setting the access token does not trigger any location services prompts and Telemetry itself will [never ask for location permission](https://github.com/mapbox/mapbox-gl-native/blob/5eda74a514964d1cac684483bafa08d458175f9a/platform/ios/src/MGLLocationManager.m#L64-L65)

**Comment by postmechanical** [comment]
@boundsj Thanks for the clarifications. Can this issue be closed then?

**Comment by boundsj** [comment]
No problem @postmechanical. I will close this now.

**Comment by cowgp** [comment]
While I understand the desire to close this - I still feel telemetry should not be initialized until an MGLMapView is alloc/init'd.  The app I work on, for some users does not initially display a map, for them it's multiple views deep before you get to one...  Yet the mere presence of MGLMapboxAccessToken makes the iOS compass icon display in the status bar.  Regardless of whether it's actually using the users location or just telemetry, that icon has historically conveyed to a user that the app is in fact using their location.   For this reason I do not believe telemetry should be tied to app launching, but be tied to the use of an MGLMapView.

**Comment by boundsj** [comment]
@cowgp Thanks so much for your feedback — we'll continue to discuss your suggestions internally (really!). The current implementation of [Telemetry](https://www.mapbox.com/telemetry) in the iOS SDK works as intended, though, so let's keep this ticket closed.
